### PR TITLE
Automatically sync stable blocks from Gutenberg to wordpress-develop on release

### PR DESCRIPTION
**Pull request by gziolo**
Replaces https://github.com/WordPress/wordpress-develop/pull/2647.

Trac ticket: https://core.trac.wordpress.org/ticket/56179.

## What problem does this PR solve?

Syncing stable blocks from the Gutenberg repository to `wordpress-develop` is currently a manual process.

The thing is, it can be automated.

That's what this PR does.

## How does this PR solve it?

The big idea is to:

1. Read all the `block.json` files from the `@wordpress/block-library` directory.
2. Pick the ones without `__experimental` restrictions – these are the stable blocks.
3. Use that list to generate PHP code that `require`s appropriate dynamic blocks PHP files.
4. Use that list to generate PHP code that calls `remove_action` with appropriate arguments in PHP unit tests.
5. Use that list yet again in Webpack config, instead of the hardcoded one.

All of this is possible thanks to https://github.com/WordPress/gutenberg/pull/40655

To make the process smooth, `require` and `remove_action` calls were extracted to separate PPH files that can be easily autogenerated and overwritten:

* `wp-includes/blocks/require-dynamic-blocks.php`
* `wp-includes/blocks/require-static-blocks.php`
* `tests/phpunit/includes/unregister-blocks-hooks.php`

## How to test?

1. Wait for the Gutenberg 13.2 release on May 11th where https://github.com/WordPress/gutenberg/pull/40655 will be included
1. Confirm all the checks are green on this PR
1. Checkout this branch
1. Run `npx grunt sync-gutenberg-packages`
1. Confirm all the stable blocks are now represented in the PHP files listed above
1. Confirm all the stable blocks were included in the build.
1. Confirm experimental blocks like `post-author-name` were excluded from both lists

cc @adamziel

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -175,6 +175,7 @@
 		"test:php": "node ./tools/local-env/scripts/docker.js run -T php composer update -W && node ./tools/local-env/scripts/docker.js run php ./vendor/bin/phpunit",
 		"test:e2e": "node ./tests/e2e/run-tests.js",
 		"test:visual": "node ./tests/visual-regression/run-tests.js",
-		"sync-gutenberg-packages": "grunt sync-gutenberg-packages --dev"
+		"sync-gutenberg-packages": "grunt sync-gutenberg-packages",
+		"postsync-gutenberg-packages": "grunt wp-packages:sync-stable-blocks && grunt build --dev"
```
Comment (gziolo) [inline package.json]: This is the simplest way to retain the ability to pass dist-tag to `sync-gutenberg-packages` and resolve the issue that blocks need to be regenerated using the dynamically loaded webpack config after the update of all packages.

The best path moving forward would be to completely rebuild the way we handle block assets in `CopyWebpackPlugin`. However, it's rather a big refactoring that would only impact this tool that is going to be executed a few times a year.

More details about the issue which led to this workaround in https://github.com/WordPress/wordpress-develop/pull/2647#issuecomment-1169879177.

Comment (adamziel) [inline package.json]: I just got confused by a PHP error – turns out this does everything we need to commit, but it doesn't prepare the local build for testing. Let's also run the regular build:

```suggestion
		"postsync-gutenberg-packages": "grunt wp-packages:sync-stable-blocks && grunt build --dev && grunt build"
```

**Review by adamziel — APPROVED** [review]
I think we're good – this is great, thank you so much for taking https://github.com/WordPress/wordpress-develop/pull/2647 across the line! With this PR, 90% of the mundane tasks required for the editor merge process are automated!

File: package.json (hunk 1 above)
Comment (gziolo) [reply package.json]: I never know which commands you need to run 😅 

We had only `grunt build --dev` before.

File: tools/release/sync-stable-blocks.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+${ staticBlockFolderNames }
+);
+`,
+	);
+
+	// wp-includes/blocks/require-blocks.php
+	console.log( 'Refreshing wp-includes/blocks/require-dynamic-blocks.php...' );
+	const dynamicBlockFileRequires = blocks
+		.filter( isDynamic )
+		.map( toDirectoryName )
```
Comment (gziolo) [inline tools/release/sync-stable-blocks.js]: I'm wondering if we should sort those directory names to ensure it's always sorted the same way. I see some subtle differences sometimes in my testing:

Screenshot 2022-07-04 at 14 23 00

Comment (adamziel) [reply tools/release/sync-stable-blocks.js]: good idea! should be as easy as `.sort()`

Comment (gziolo) [reply tools/release/sync-stable-blocks.js]: Done with [ace61e4](https://github.com/WordPress/wordpress-develop/pull/2940/commits/ace61e468806c2a96e7ff4a0769ed028fdacdac5).

**Comment by gziolo** [comment]
I landed most of the changes with https://core.trac.wordpress.org/changeset/53688. The last remaining piece is the changes to the weback config that requires updating core blocks first.

**Comment by azaozz** [comment]
Any particular reason the auto-generated PHP files are committed to trunk? Think it was decided to not do that again after the problems with `script-loader-packages.php`. 

Also, these files are supposed to be overwritten every time WP is build, right? So why commit the potentially old versions of them :)

**Comment by gziolo** [comment]
With https://core.trac.wordpress.org/changeset/54308 all changes are now included in WordPress core.

**Pull request closed without merging**